tests/backend_test_base.py

```diff
@@ -83,47 +93,74 @@ def run_test_case(self, feed_dict, input_names_with_port, output_names_with_port
+            concrete_func = concrete_func.get_concrete_function()
+            graph_def = from_function(concrete_func,
+                                      input_names=list(feed_dict.keys()), output_names=output_names_with_port)
+        else:
+            #
+            # use graph to execute the tensorflow func
+            #
+            with tf_session() as sess:
+                tf.set_random_seed(1)
+                input_list = []
+                for k, v in clean_feed_dict.items():
+                    input_list.append(tf_placeholder(name=k, shape=v.shape, dtype=tf.as_dtype(v.dtype)))
+                func(*input_list)
+                variables_lib.global_variables_initializer().run()
+                tf.tables_initializer().run()
+                output_dict = []
+                for out_name in output_names_with_port:
+                    output_dict.append(sess.graph.get_tensor_by_name(out_name))
+                expected = sess.run(output_dict, feed_dict=feed_dict)
+                graph_def = freeze_session(sess,
+                                           input_names=list(feed_dict.keys()),
+                                           output_names=output_names_with_port)
+
+            tf_reset_default_graph()
+            with tf_session() as sess:
+                tf.import_graph_def(graph_def, name='')
+                input_tensors = {i: sess.graph.get_tensor_by_name(i) for i in list(feed_dict.keys())}
+                output_tensors = {i: sess.graph.get_tensor_by_name(i) for i in output_names_with_port}
+                graph_def = tf_optimize(input_tensors, output_tensors, graph_def, fold_constant=constant_fold)
+
+        tf_reset_default_graph()
+        with tf_session() as sess:
             tf.import_graph_def(graph_def, name='')
 
-        with tf.Session() as sess:
-            tf.tables_initializer().run()
-            variables_lib.global_variables_initializer().run()
-            output_dict = []
-            for out_name in output_names_with_port:
-                output_dict.append(sess.graph.get_tensor_by_name(out_name))
-            expected = sess.run(output_dict, feed_dict=feed_dict)
-
-        if self.config.is_debug_mode:
-            if not os.path.exists(self.test_data_directory):
-                os.makedirs(self.test_data_directory)
-            model_path = os.path.join(self.test_data_directory, self._testMethodName + "_original.pb")
-            utils.save_protobuf(model_path, sess.graph_def)
-            self.logger.debug("created file %s", model_path)
-
-        graph_def = tf_optimize(input_names_with_port, output_names_with_port,
-                                sess.graph_def, constant_fold)
-
-        if self.config.is_debug_mode:
-            model_path = os.path.join(self.test_data_directory, self._testMethodName + "_after_tf_optimize.pb")
-            utils.save_protobuf(model_path, graph_def)
-            self.logger.debug("created file  %s", model_path)
```

Comment:
This might be useful to keep around.

Reply:
yeap, it is still there but moved a few lines down.